### PR TITLE
Replace libraries and fix several bugs

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Rework trigger system (#303)
  - Create class `TriggerEvent` and `TriggerEventHandler`
    - `TriggerEvent`: This is an event class for trigger
    - `TriggerEventHandler`: This is an event handler class for trigger
  - Fix the logic of cards
    - 'Young Priestess' (EX1_004): Change `Effects::HealthN(2)` to `Effects::HealthN(1)`
    - 'Conceal' (EX1_128): Add self condition `SelfCondition::HasNotStealth()`
      - `SelfCondition::HasNotStealth()`: SelfCondition wrapper for checking the entity has not stealth
      - References: https://hearthstone.gamepedia.com/Conceal
- Replace library 'Clara' with 'Lyra' (#389)
- Change unit test framework: googletest -> doctest (#392)
- Create structure `CardDefs` (#393)
  - Create class `CardDef` and `CardDefs`
    - `CardDef`: This class stores the card data such as powers and play requirements
      - Variable list: power, playRequirements, chooseCardIDs, entourages, questProgressTotal
    - `CardDefs`: This class manages a list of CardDef data
- Analyze the causes of intermittent unit test failure (#397)
  - 'Brightwing' (EX1_189) and 'Lifeweaver' (DAL_355)
- Rework code that processes ability 'Choose One' (#398)
  - 'Wardruid Loti' (TRL_343) and 'Rising Winds' (YOD_001)
- Fix the logic of 'Cenarius' (EX1_573), 'Onyxia' (EX1_562)
  - References: https://hearthstone.gamepedia.com/Cenarius 
  - References: https://hearthstone.gamepedia.com/Onyxia

### CODE REVIEW

File: Includes/Rosetta/Cards/CardDef.hpp [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+//!
+class CardDef
+{
+ public:
+    //! Default constructor.
+    CardDef() = default;
+
+    //! Constructs card def with given \p _power.
+    //! \param _power The power data.
+    CardDef(Power _power);
```
Comment (utilForever) [inline Includes/Rosetta/Cards/CardDef.hpp]: Codacy Issue found: [Class 'CardDef' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5011091)

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Clones removed
==============
+ Includes/Rosetta/CardSets/CoreCardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/CountTaskTests.cpp  -2
+ Includes/Rosetta/CardSets/DalaranCardsGen.hpp  -1
+ Sources/Rosetta/CardSets/CoreCardsGen.cpp  -1
+ Includes/Rosetta/CardSets/HoFCardsGen.hpp  -1
+ Tests/UnitTests/Views/BoardTests.cpp  -1
+ Includes/Rosetta/CardSets/DragonsCardsGen.hpp  -1
+ Extensions/RosettaConsole/Console.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/DamageTaskTests.cpp  -2
+ Includes/Rosetta/CardSets/YoDCardsGen.hpp  -1
+ Tests/UnitTests/Views/BoardRefViewTests.cpp  -2
+ Sources/Rosetta/CardSets/HoFCardsGen.cpp  -1
+ Includes/Rosetta/CardSets/Expert1CardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/GetEventNumberTaskTests.cpp  -2
+ Tests/UnitTests/Enchants/PlayerAuraEffectsTests.cpp  -1
+ Extensions/RosettaTool/main.cpp  -1
+ Tests/UnitTests/Games/GameTests.cpp  -1
+ Includes/Rosetta/CardSets/UldumCardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/DestroyTaskTests.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5011091&bid=16528064)

File: Includes/Rosetta/Managers/TriggerEventHandler.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+{
+ public:
+    using Func = std::function<void(Entity*)>;
+
+    //! Default constructor.
+    TriggerEventHandler();
+
+    //! Constructs trigger event handler with given \p func.
+    //! \param func A functor to run.
+    TriggerEventHandler(Func func);
```
Comment (utilForever) [inline Includes/Rosetta/Managers/TriggerEventHandler.hpp]: Codacy Issue found: [Class 'TriggerEventHandler' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5011091)

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ Includes/Rosetta/CardSets/DragonsCardsGen.hpp  -1
+ Extensions/RosettaTool/main.cpp  -1
+ Sources/Rosetta/CardSets/CoreCardsGen.cpp  -1
+ Includes/Rosetta/CardSets/CoreCardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/DestroyTaskTests.cpp  -2
+ Tests/UnitTests/Enchants/PlayerAuraEffectsTests.cpp  -1
+ Tests/UnitTests/Games/GameTests.cpp  -1
+ Includes/Rosetta/CardSets/Expert1CardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/CountTaskTests.cpp  -2
+ Extensions/RosettaConsole/Console.hpp  -1
+ Includes/Rosetta/CardSets/DalaranCardsGen.hpp  -1
+ Includes/Rosetta/CardSets/HoFCardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/DamageTaskTests.cpp  -2
+ Sources/Rosetta/CardSets/HoFCardsGen.cpp  -1
+ Includes/Rosetta/CardSets/UldumCardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/GetEventNumberTaskTests.cpp  -2
+ Tests/UnitTests/Views/BoardTests.cpp  -1
+ Includes/Rosetta/CardSets/YoDCardsGen.hpp  -1
+ Tests/UnitTests/Views/BoardRefViewTests.cpp  -2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5011091&bid=16528064)

File: Sources/Rosetta/Tasks/SimpleTasks/AddEnchantmentTask.cpp [hunk 3]
```diff
@@ -10,13 +10,17 @@
...
+
 namespace RosettaStone::SimpleTasks
 {
-AddEnchantmentTask::AddEnchantmentTask(const std::string& cardID,
-                                       EntityType entityType, bool useScriptTag)
+AddEnchantmentTask::AddEnchantmentTask(
+    const std::string& cardID, EntityType entityType, bool useScriptTag,
+    std::optional<SelfCondition> selfCondition)
     : ITask(entityType),
       m_enchantmentCard(Cards::FindCardByID(cardID)),
```
Comment (CreatorSeraph) [inline Sources/Rosetta/Tasks/SimpleTasks/AddEnchantmentTask.cpp]: cardID를 std::string_view&로 받아도 될 것 같습니다.

Comment (utilForever) [reply Sources/Rosetta/Tasks/SimpleTasks/AddEnchantmentTask.cpp]: 수정했습니다.

**Comment by utilForever** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones removed
==============
+ Tests/UnitTests/Views/BoardTests.cpp  -1
+ Extensions/RosettaTool/main.cpp  -1
+ Includes/Rosetta/CardSets/HoFCardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/GetEventNumberTaskTests.cpp  -2
+ Tests/UnitTests/Views/BoardRefViewTests.cpp  -2
+ Includes/Rosetta/CardSets/DragonsCardsGen.hpp  -1
+ Includes/Rosetta/CardSets/Expert1CardsGen.hpp  -1
+ Includes/Rosetta/CardSets/UldumCardsGen.hpp  -1
+ Sources/Rosetta/CardSets/HoFCardsGen.cpp  -1
+ Extensions/RosettaConsole/Console.hpp  -1
+ Includes/Rosetta/CardSets/DalaranCardsGen.hpp  -1
+ Includes/Rosetta/CardSets/CoreCardsGen.hpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/DestroyTaskTests.cpp  -2
+ Tests/UnitTests/Tasks/SimpleTasks/DamageTaskTests.cpp  -2
+ Includes/Rosetta/CardSets/YoDCardsGen.hpp  -1
+ Tests/UnitTests/Enchants/PlayerAuraEffectsTests.cpp  -1
+ Tests/UnitTests/Tasks/SimpleTasks/CountTaskTests.cpp  -2
+ Tests/UnitTests/Games/GameTests.cpp  -1
+ Sources/Rosetta/CardSets/CoreCardsGen.cpp  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/utilForever/RosettaStone/pullRequest?prid=5011091&bid=16528064)

**Pull request merged**